### PR TITLE
Fix undefined user and enable atomic flake installs

### PR DESCRIPTION
**Pull request by worldofgeese**
A user running `ansible-playbook playbook.yml -i inventory --ask-become-pass` will encounter the error `ansible_user is undefined`. This PR changes `ansible_user` to a `$USER` lookup.

Additionally, the single quotes inside double quotes for the Nix feature flags makes them unreadable by Nix: they will be ignored given any Nix invocation. This also removes those single quotes.

### CODE REVIEW

**Comment by worldofgeese** [comment]
I've also pushed a fix to #75, expanding the scope of this PR.

**Review by nre-ableton — COMMENTED** [review]
First, thanks for submitting this PR! Before we can merge it, we'll need you to sign our company's [CLA](https://ableton.github.io/cla/).

In regards to the PR itself, the idea here seems sound. After doing some tests, it seems that you are indeed correct and the single quotes cause nix to fail with a parse error. We don't use nix-flakes ourselves in our infrastructure, so this one must have slipped by us. :sweat_smile: 

I would like to figure out how to make a molecule test to actually verify that the nix installation works, to avoid regressions like this in the future. Would you mind to enable the "Allow contributions from maintainers" option on this branch? I will try to push a fixup commit or two.

**Comment by nre-ableton** [comment]
BTW, I just got an email confirming the CLA. Thanks! :+1:

File: tasks/main.yml
Comment (nre-ableton) [reply tasks/main.yml:78]: Actually, please ignore this comment. I think that more work must be done in this task. I'm working on a solution now.

File: tasks/main.yml
Comment (worldofgeese) [reply tasks/main.yml:51]: > So if I understand you correctly from your comments in https://github.com/Ableton/ansible-role-nix/issues/75, you want to run this role on a system where nix has already been installed

I am testing with Multipass which allows me to spin up fresh Ubuntu VMs for every run. On brand new Ubuntu VMs without Nix being installed prior, the flake install skips. Are you able to test this with Multipass too?

You're right about losing idempotency however, from memory without checking, I believe the Nix install script gracefully detects pre-existing installs when run and will skip itself.

File: tasks/main.yml
Comment (worldofgeese) [reply tasks/main.yml:78]: If we can resolve the trouble I'm having with the `when: nix_directory is changed` condition on a Nix-naive Ubuntu VM, I think your suggestion is good! But I'd like to hear your own results first before making this commit.

**Comment by nre-ableton** [comment]
@worldofgeese it seems that I don't have permission to push to your branch (unless I am missing something), could you please enable the option to allow maintainers to make edits?

**Comment by worldofgeese** [comment]
I sent an invite to be a collaborator which should allow access?

File: defaults/main.yml
Comment (nre-ableton) [reply defaults/main.yml]: Sorry, I fat-fingered this and missed the space after the comma. I pushed a fixup, a3b3b3e433835649ce18383d18d2e5c85f2c91d0.

File: tasks/main.yml
Comment (nre-ableton) [reply tasks/main.yml:51]: We don't use multipass, but we have an internal vCenter service that I can use for the same purpose. With fresh Ubuntu VMs, we see nix being installed successfully. Also, the molecule tests should also demonstrate that nix can be installed on a nix-less machine. Regarding the idempotency, this change does break the molecule tests.

I've pushed a fixup to revert this, f56a6b7.

So I still think that this condition needs to stay. However, I now agree that the second condition should drop the `nix_directory is changed` bit, but let's discuss that in the other thread.

File: tasks/main.yml
Comment (nre-ableton) [reply tasks/main.yml]: Ok, so as mentioned above, you are right, the `nix_directory is changed` bit needs to go. What we _should_ do is instead see if we've already written the `nix-command flakes` bit in the config file (or if the config file even exists). Ideally, it would be nice if we could run a nix command to tell us if flakes have been configured, but I'm not aware of one (and I'm not very knowledgeable about nix, so please let me know if I'm missing something here). I tried `nix-env --query --installed`, but it doesn't list flakes there, so I guess it isn't treated like other packages.

Anyways, I've implemented these checks in https://github.com/Ableton/ansible-role-nix/pull/74/commits/7d5d60786fb574855d2b450793d4a293644a6e36. With this change, we should only need the `when: nix_flakes` condition on that `block`.

**Comment by nre-ableton** [comment]
> I sent an invite to be a collaborator which should allow access?

Yep, got it, thanks.

Would you mind if I rewrote the git history and force-pushed to this branch? This would collapse all of my fixups and make the branch a bit easier to review.

**Comment by worldofgeese** [comment]
Go for it! My (git repository) house is your house ☺️

**Review by nre-ableton — APPROVED** [review]
@worldofgeese Ok, I've squashed/pushed, and also added a version bump (which means that a new version will be published after this PR is merged). Everything looks fine on my end, but it would be nice if you could test these changes again with your Multipass setup just to make sure that the `nix_directory is changed` condition around the "Install nix" block is safe.

Assuming that these changes still work for you, I'd be happy to merge this PR and make a release. :+1:

**Comment by worldofgeese** [comment]
I launched a brand new Multipass Ubuntu 20.04.4 LTS VM, installed Ansible then ran
```
ansible-galaxy install -r requirements.yml --force
ansible-playbook playbook.yml -i inventory --ask-become-pass -e '{"nix_flakes": true}'
```
in from a freshly cloned https://github.com/worldofgeese/provision-ubuntu-on-wsl2.

I can confirm the run completes with no complaints! One remaining annoyance is that the user's shell must be exited after running the playbook before Nix is available on the user's command line.

Would it be useful to source `if [ -e {{ nix_user_details.home }}/.nix-profile/etc/profile.d/nix.sh ]; then . /home/{{ nix_user_details.home }}/.nix-profile/etc/profile.d/nix.sh; fi` at the end of the tasks block?

**Comment by nre-ableton** [comment]
Yeah, unfortunately, I don't think there's any way to get around the fact that the user must restart their shell. However, we can at least document this, which I have done in a6e6b39.

@worldofgeese I'm going to go ahead and merge this PR. Does this mean that we can also close #75?

**Comment by nre-ableton** [comment]
Ok, I've just merged #76, which was basically this branch but with a different origin, so this PR is no longer needed. Version 0.0.3 of this role was published a few minutes ago. Thanks a lot @worldofgeese for your contribution! :tada:

**Pull request closed without merging**
